wordpress-mobile/WordPress-iOS
# Move uploading media to WP Media Library to `MediaImportService`

There is the usual accessing managed object in a wrong thread issue (I'll highlight a few in the PR comments) in the implementation of uploading a media asset to the WP Media Library.

This PR rewrites (large chunk of the original code is still reused) the uploading logic to make the Core Data operations safer, by adopting the new writer API.

To help with the review, I'll quickly described what happens during uploading a media asset to WP media library. The entry point is [the `MediaCoordinator.addMedia` function](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaCoordinator.swift#L161), which can be called during editing a post or simply uploading an photo (on device or from the free online photo library) on the "Media" screen. This function first imports the photo into the app's sandbox, using [`MediaImportService`](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaImportService.swift#L49). Then it generates a thumbnail using [`MediaThumbnailService`](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaThumbnailService.swift#L47). At the same time, it starts to upload the imported media to the WP media library, using [`MediaService`](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaService.m#L199).

The refactor does not change the above logic, but moves some code around, makes the Core Data operations safer, and changed a few call sites to adopt to the new API.

## Test instructions

Verify all the places that can upload a photo to WP media library. I've tested uploading a new site icon, uploading a photo to the Media Library on "Media" screen, uploading a photo to a post in the post editor.

## Regression Notes

1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
What described in the "Test instructions"

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:
